test/common/testlib.py

```diff
@@ -1095,7 +1032,131 @@ def copy_cores(self, title, label=None):
+            machine.start()
+
+        # Now wait for the other machines to be up
+        for key in self.machines.keys():
+            machine = self.machines[key]
+            machine.wait_boot()
+            address = provision[key].get("address")
+            if address is not None:
+                machine.set_address(address)
+            dns = provision[key].get("dns")
+            if address or dns:
+                machine.set_dns(dns)
+            dhcp = provision[key].get("dhcp", False)
+            if dhcp:
+                machine.dhcp_server()
+
+        if self.machine:
+            self.journal_start = self.machine.journal_cursor()
+            self.browser = self.new_browser()
+        self.tmpdir = tempfile.mkdtemp()
+
+        def sitter():
+            if opts.sit and not self.checkSuccess():
+                if self._outcome:
+                    [traceback.print_exception(*e[1]) for e in self._outcome.errors if e[1]]
+                else:
+                    self.currentResult.printErrors()
+                sit(self.machines)
+        self.addCleanup(sitter)
+
+        def intercept():
+            if not self.checkSuccess():
+                self.snapshot("FAIL")
+                self.copy_js_log("FAIL")
+                self.copy_journal("FAIL")
+                self.copy_cores("FAIL")
+        self.addCleanup(intercept)
+
+    def tearDown(self):
+        if self.checkSuccess() and self.machine.ssh_reachable:
+            self.check_journal_messages()
+            self.check_browser_errors()
+        shutil.rmtree(self.tmpdir)
+
+
+class PersistentMachineCase(BaseCase):
+    @classmethod
+    def setUpClass(cls):
+        if cls.provision is not None:
+            raise unittest.SkipTest("Cannot provision machines on a PersistentMachineCase")
+
+        cls.machine = None
+        cls.browser = None
+        cls.machine_options = {}
+
+        # First create all machines, wait for them later
+        cls.machine = cls.new_machine(cls, cleanup=False, **cls.machine_options)
+        if opts.trace:
+            print("Starting {0} {1}".format('machine1', cls.machine.label))
+        cls.machine.start()
+        cls.machine.wait_boot()
+        cls.tmpdir = tempfile.mkdtemp()
+
+    def setUp(self):
+        self.journal_start = self.machine.journal_cursor()
+        self.browser = self.new_browser()
+
+        def sitter():
+            if opts.sit and not self.checkSuccess():
+                if self._outcome:
+                    [traceback.print_exception(*e[1]) for e in self._outcome.errors if e[1]]
+                else:
+                    self.currentResult.printErrors()
+                sit(self.machines)
+
+        def intercept():
+            if not self.checkSuccess():
+                self.snapshot("FAIL")
+                self.copy_js_log("FAIL")
+                self.copy_journal("FAIL")
+                self.copy_cores("FAIL")
+
+        self.addCleanup(sitter)
+        self.addCleanup(intercept)
```

Comment:
This all looks appropriate for the BaseCase, and the MachineCase can then just do the additional stuff?